File: src/main/java/io/github/shaksternano/mediamanipulator/media/graphics/GraphicsUtil.java
package io.github.shaksternano.mediamanipulator.media.graphics;

import io.github.shaksternano.mediamanipulator.io.FileUtil;
import io.github.shaksternano.mediamanipulator.media.graphics.drawable.Drawable;

import java.awt.*;
import java.io.IOException;
import java.io.ObjectInputStream;

public class GraphicsUtil {

    public static int fontFitWidth(int maxWidth, Drawable text, Graphics2D graphics) {
        Font font = graphics.getFont();
        int textWidth = text.getWidth(graphics);
        while (textWidth > maxWidth) {
            float sizeRatio = (float) textWidth / maxWidth;
            font = font.deriveFont(font.getSize() - sizeRatio);
            graphics.setFont(font);
            textWidth = text.getWidth(graphics);
        }

        return textWidth;
    }

    public static int fontFitHeight(int maxHeight, Drawable text, Graphics2D graphics) {
        Font font = graphics.getFont();
        int textHeight = text.getHeight(graphics);
        while (textHeight > maxHeight) {
            float sizeRatio = (float) textHeight / maxHeight;
            font = font.deriveFont(font.getSize() - sizeRatio);
            graphics.setFont(font);
            textHeight = text.getHeight(graphics);
        }

        return textHeight;
    }

    public static Shape loadShape(String resourcePath) throws IOException {
        try (ObjectInputStream input = new ObjectInputStream(FileUtil.getResourceInRootPackage(resourcePath))) {
            Object parsedObject = input.readObject();
            if (parsedObject instanceof Shape shape) {
                return shape;
            } else {
                throw new IOException("Failed to load shape file under \"" + resourcePath + "\"! The parsed object is not a shape!");
            }
        } catch (ClassNotFoundException e) {
            throw new IOException("Failed to load shape file under \"" + resourcePath + "\"!", e);
        }
    }
}

File: core/src/main/kotlin/io/github/shaksternano/borgar/core/graphics/GraphicsUtil.kt
package io.github.shaksternano.borgar.core.graphics

import io.github.shaksternano.borgar.core.graphics.drawable.Drawable
import java.awt.Graphics2D
import java.awt.Rectangle
import java.awt.RenderingHints

fun Graphics2D.configureTextDrawQuality() {
    setRenderingHint(
        RenderingHints.KEY_FRACTIONALMETRICS,
        RenderingHints.VALUE_FRACTIONALMETRICS_ON
    )

    setRenderingHint(
        RenderingHints.KEY_TEXT_ANTIALIASING,
        RenderingHints.VALUE_TEXT_ANTIALIAS_LCD_HRGB
    )

    setRenderingHint(
        RenderingHints.KEY_ANTIALIASING,
        RenderingHints.VALUE_ANTIALIAS_ON

    )

    setRenderingHint(
        RenderingHints.KEY_RENDERING,
        RenderingHints.VALUE_RENDER_QUALITY
    )
}

fun Graphics2D.fitFontWidth(maxWidth: Int, text: Drawable): Int {
    var font = this.font
    var textWidth = text.getWidth(this)
    while (textWidth > maxWidth) {
        val sizeRatio = textWidth.toFloat() / maxWidth
        font = font.deriveFont(font.size - sizeRatio)
        this.font = font
        textWidth = text.getWidth(this)
    }
    return textWidth
}

fun Graphics2D.fitFontHeight(maxHeight: Int, text: Drawable): Int {
    var font = this.font
    var textHeight = text.getHeight(this)
    while (textHeight > maxHeight) {
        val sizeRatio = textHeight.toFloat() / maxHeight
        font = font.deriveFont(font.size - sizeRatio)
        this.font = font
        textHeight = text.getHeight(this)
    }
    return textHeight
}

fun Graphics2D.fillRect(rectangle: Rectangle) {
    fillRect(rectangle.x, rectangle.y, rectangle.width, rectangle.height)
}

data class Position(val x: Int, val y: Int)
